### ISSUE TITLE
\chapterformat has no effect if default language is Russian

### ISSUE THREAD

**Issue by intractabilis** [reutenauer/polyglossia#444]
You cannot redefine chapter heading using `\chapterformat` from KOMA-script, if the default language is set to `russian`.

MWE:
```LaTeX
\documentclass{scrbook}

\usepackage{polyglossia}
\usepackage{fontspec}

\setdefaultlanguage{russian}

\setmainfont{NewComputerModern}
\newfontfamily{\cyrillicfont}{NewComputerModern}
\setsansfont{NewComputerModernSans}
\newfontfamily{\cyrillicfontsf}{NewComputerModernSans}
\setmonofont{NewComputerModernMono}
\newfontfamily{\cyrillicfonttt}{NewComputerModernMono}

\renewcommand*{\chapterformat}{%
  \mbox{\chapappifchapterprefix{\nobreakspace} some text \thechapter
  \autodot\IfUsePrefixLine{}{\enskip}}}

\begin{document}
\chapter{Chapter}

\end{document}
```

Expected output: `some text 1 Chapter`.
Instead we get: `1. Chapter`.

**Comment by jspitz** [comment]
Yes, gloss-russian is redefining `\chapterformat`, I think we need an option to opt out of this behavior (as we have for Hungarian).

**Comment by intractabilis** [comment]
So far I found a workaround: redefine `\chapterformat` after every `\selectlanguage` in the `document` environment.

**Comment by jspitz** [comment]
> So far I found a workaround: redefine `\chapterformat` after every `\selectlanguage` in the `document` environment.

You just need to do once:
```tex
\makeatletter
\addto\blockextras@russian{%
\renewcommand*{\chapterformat}{%
  \mbox{\chapappifchapterprefix{\nobreakspace} some text \thechapter
  \autodot\IfUsePrefixLine{}{\enskip}}}
}
\makeatother
```

(you might also need to `\addto\noextras@russian`)

> I think it would have been less disruptive, and in sync with the KOMA-Script principles, if `gloss-russian` had defined only `\thechapter` and `\autodot`, which are the parts of the default `\chapterformat`.

I don't know the backgrounds of this chapter reformatting. It has been there since commit one of polyglossia.

**Comment by jspitz** [comment]
OK, I had a closer look and think I understand. The only thing that this whole code attempts is to add a period after all section headings in Russian, right? (the spacing is not really changed). If so, I agree that this code takes the sledgehammer to crack the nut.

Could you please try if
https://github.com/jspitz/polyglossia/blob/master/tex/gloss-russian.ldf

is identically in terms of output? This is much more to the point, and it also works with your modification.

**Comment by intractabilis** [comment]
> OK, I had a closer look and think I understand. The only thing that this whole code attempts is to add a period after all section headings in Russian, right? (the spacing is not really changed). If so, I agree that this code takes the sledgehammer to crack the nut.

Haha, in Russian we say "use a cannon to shoot sparrows". I am not very TeX savvy to say what this code is doing. However, if it enforces a period after all section headings, it is wrong. KOMA Script defaults is exactly what is needed.

> is identically in terms of output? This is much more to the point, and it also works with your modification.

Enforcing a period is incorrect. My whole story began from the point where I was trying to design a chapter heading and found out that the chapter number always has a period after it:
![изображение](https://user-images.githubusercontent.com/45959195/94219317-640a3d00-fe9b-11ea-82a7-fc0a6e083bed.png)

Since I couldn't reset it, I posted this issue. Now this fix makes it even worse: even the redefinition of `chapterformat` doesn't work. So the fix is doing the opposite of what is needed.

The bottom line: please remove the enforcement of the period completely. If KOMA Script is intact, its defaults is exactly what is needed.

**Comment by intractabilis** [comment]
> Could you please try if
> https://github.com/jspitz/polyglossia/blob/master/tex/gloss-russian.ldf

Btw, thanks for adding `indentfirst = true` to defaults for Russian in that fix.

**Comment by jspitz** [comment]
I am not familiar with Russian typesetting conventions, but it has been a deliberate change since day 1 of poylglossia, so I won't break that just like that. @MaksimZh  can you advise?

**Comment by jspitz** [comment]
> Now this fix makes it even worse: even the redefinition of `chapterformat` doesn't work. So the fix is doing the opposite of what is needed.

Hm that's strange. This works for me with your test case.

**Comment by intractabilis** [comment]
> I am not familiar with Russian typesetting conventions, but it has been a deliberate change since day 1 of poylglossia, so I won't break that just like that. @MaksimZh can you advise?

There is a period by convention and **by default KOMA Script adds it** (in fact conventions in Germany and in Russia are quite similar). There is no need in extra enforcement. It just confuses a user if she reads in the KOMA Script documentation that `\documentclass[numbers = noenddot]{scrbook}` should remove the dot, but suddenly polyglossia decided to tweak that. It costs hours of search for users why suddenly to their utter surprise `numbers = noenddot` stops working.

**Comment by intractabilis** [comment]
> > Now this fix makes it even worse: even the redefinition of `chapterformat` doesn't work. So the fix is doing the opposite of what is needed.
> 
> Hm that's strange. This works for me with your test case.

MWE is below. It produces
![изображение](https://user-images.githubusercontent.com/45959195/94241955-35f02180-feca-11ea-960c-fa4c5f1c2a53.png)

There is a dot after the chapter number, which shouldn't be there if `numbers = noenddot` is used. With your new change adding
```LaTeX
\addto\blockextras@russian{%
\renewcommand*{\chapterformat}{%
  \mbox{\chapappifchapterprefix{\nobreakspace}\thechapter
  \autodot\IfUsePrefixLine{}{\enskip}}}
}
```
doesn't help anymore. Though I have to admit, removing `\autodot` will help, IMHO it is very confusing for users. We are solving a non-existent problem here. KOMA Script already takes good care for the dot.

Promised MWE per se:

```LaTeX
\documentclass[12pt, letterpaper, numbers = noenddot]{scrbook}

\usepackage[autopunct]{csquotes}
\usepackage{polyglossia}
\usepackage{tikz}

\setdefaultlanguage[babelshorthands = true]{russian}
\SetLanguageKeys{russian}{
    hyphenmins  = {2, 2}
  , indentfirst = true
  }

% it's easier to notice with Garamond
\setmainfont{EB Garamond}[
    Numbers    = {Proportional, Uppercase}
  , RawFeature = {+ss6}
  ]
\newfontfamily\cyrillicfont[
    Numbers           = {Proportional, Uppercase}
  , RawFeature        = {+ss6}
  ]{EB Garamond}

\setsansfont{FreeSans}
\newfontfamily\cyrillicfontsf{FreeSans}
\setmonofont{FreeMono}
\newfontfamily\cyrillicfontmono{FreeMono}

\KOMAoptions{fontsize = 14, BCOR = 8.25mm}
\KOMAoptions{DIV = 14}

\pghyphenation{russian}{
    фран-цу-жен-кою
}

\definecolor{chaptercolor}{RGB}{205, 130, 130}
\renewcommand\raggedchapter{\raggedleft}
\setkomafont{chapter}{\Huge}
\setkomafont{chapterprefix}{\large}
\newkomafont{chapternumber}{\mdseries\fontsize{50pt}{60pt}\selectfont}

\tikzset{
    headings/base/.style = {
        outer sep = 0pt
      , inner sep = 5pt
      }
  , headings/chapterbackground/.style = {
       headings/base
      , shade
      , left color = white
      , right color = chaptercolor
      }
  , headings/chapapp/.style = {
        headings/base
      , text = chaptercolor
      , font = \usekomafont{chapterprefix}
      }
  , headings/chapternumber/.style= {
        headings/base
      , text = chaptercolor
      , font = \usekomafont{chapternumber}
      }
  , headings/chapterline/.style = {
        chaptercolor
      , line width = 2pt
      }
  }

\makeatletter
\renewcommand*\chapterlinesformat[3]{%
    \Ifstr{#1}{chapter}{%
        \begin{tikzpicture}[baseline = (title.base)]
            \node[headings/chapterbackground](title){%
                \parbox[t]
                {\dimexpr\textwidth-2\pgfkeysvalueof{/pgf/inner xsep}\relax}
                {\raggedchapter #3}%
            };
            \node[headings/chapapp, anchor = south east]
                at (title.north east){\Ifstr{#2}{}{}{\chapapp}\strut};
            \useasboundingbox
                (current bounding box.north west)
                rectangle
                ([yshift = -10pt] current bounding box.south east);
            \draw[headings/chapterline]
                (current bounding box.south east)
                ++(+.5\pgflinewidth, 0) -- +(0, \paperheight);
            \node[
                anchor = base west
              , headings/chapternumber
              ] at ([xshift = 10pt]title.base -| current bounding box.east){#2};
        \end{tikzpicture}
        \par
    }{%
    \@hangfrom{#2}{#3}
  }%
}
\makeatother

\begin{document}

\chapter{Всё смешалось…}

Все счастливые семьи похожи друг на друга, каждая несчастливая семья
несчастлива по-своему.

Всё смешалось в доме Облонских. Жена узнала, что муж был в связи с бывшею в их доме
француженкою-гувернанткой, и объявила мужу, что не может жить с ним в одном доме. Положение это
продолжалось уже третий день и мучительно чувствовалось и самими супругами, и всеми членами семьи, и
домочадцами. Все члены семьи и домочадцы чувствовали, что нет смысла в их сожительстве и что на
каждом постоялом дворе случайно сошедшиеся люди более связаны между собой, чем они, члены семьи и
домочадцы Облонских. Жена не выходила из своих комнат, мужа третий день не было дома. Дети бегали по
всему дому, как потерянные; англичанка поссорилась с экономкой и написала записку приятельнице,
прося приискать ей новое место; повар ушел вчера со двора, во время самого обеда; чёрная кухарка и
кучер просили расчёта.

На третий день после ссоры князь Степан Аркадьич Облонский~— Стива, как его звали в свете,~— в
обычный час, то есть в восемь часов утра, проснулся не в спальне жены, а в своем кабинете, на
сафьянном диване. Он повернул свое полное, выхоленное тело на пружинах дивана, как бы желая опять
заснуть надолго, с другой стороны крепко обнял подушку и прижался к ней щекой; но вдруг вскочил, сел
на диван и открыл глаза.

\textquote{Да, да, как это было?~— думал он, вспоминая сон.~— Да, как это было? Да! Алабин
давал обед в Дармштадте; нет, не в Дармштадте, а что-то американское. Да, но там Дармштадт был в
Америке. Да, Алабин давал обед на стеклянных столах, да,~— и столы пели: Il mio
tesoro\footnote{Моё сокровище (\textit{итал.}).} и не Il mio tesoro, а что-то лучше, и какие-то
маленькие графинчики, и они же женщины},~— вспоминал он.

\end{document}
```

**Comment by jspitz** [comment]
Yes, right, thus I propose an opt-out option to turn this off (see https://github.com/reutenauer/polyglossia/issues/444#issuecomment-696507637)

**Comment by jspitz** [comment]
But let me rephrase my question: Apart from your attempt to change it, does the output with my change look the same than before or not?

**Comment by intractabilis** [comment]
> Yes, right, thus I propose an opt-out option to turn this off (see [#444 (comment)](https://github.com/reutenauer/polyglossia/issues/444#issuecomment-696507637))

Opt-out the redefinition of `\chapterformat`? What issue does this redefinition solve in the first place? From your words it just adds a period after a chapter number. Which already the default behavior of KOMA Script. Therefore, the only result of it is extra maintenance headache for you.

> But let me rephrase my question: Apart from your attempt to change it, does the output with my change look the same than before or not?

It doesn't. It looks like this:
![изображение](https://user-images.githubusercontent.com/45959195/94249694-1c080c00-fed5-11ea-999f-9fbab3d947d6.png)

Before it looked like this:
![изображение](https://user-images.githubusercontent.com/45959195/94249762-32ae6300-fed5-11ea-9a39-704d738db1c1.png)

**Comment by jspitz** [comment]
> Opt-out the redefinition of `\chapterformat`? 

Opt-out the redefinition of `\autodot` (or any fiddling with the format of chapter headings).

> What issue does this redefinition solve in the first place? From your words it just adds a period after a chapter number. Which already the default behavior of KOMA Script. 

No, KOMA-Script's default is no dot after headings with Arabic numbers (`numbers=auto`)

> > But let me rephrase my question: Apart from your attempt to change it, does the output with my change look the same than before or not?
> 
> It doesn't. It looks like this:

This contains your redefinitions. I was asking about default behavior (without any `\chapterformat` fiddling)

**Comment by intractabilis** [comment]
> No, KOMA-Script's default is no dot after headings with Arabic numbers (`numbers=auto`)

Thank you for correcting me. You didn't answer the question what issue does the redefinition solve for Russian language. Though, I feel I should disengage from the discussion. I stated my opinion and know the workarounds. Convincing you to do less work is not very constructive. You can close the issue, I am fine with any decision or no changes at all.

> I was asking about default behavior (without any `\chapterformat` fiddling)

Yes, it is the same and both are wrong.

**Comment by jspitz** [comment]
It's not about me having a fixed position. As I said, I know next to nothing about how it is supposed to look like. But I am reluctant to change a behavior of `polyglossia` that has been there since the beginning and that other Russian-speaking users have invested energy to implement. The more so since this will change the output of existing documents behind the back of users (breaks backwards compatibility). That's why I am asking.

In any case, would it work for you if you could set (e.g.)
```tex
\setmainlanguage[forceheadingpunctuation=false]{russian}
```
to opt-out of polyglossia's tweaking and chime in with your own modifications?

**Comment by intractabilis** [comment]
> The more so since this will change the output of existing documents behind the back of users (breaks backwards compatibility).

It won't change the output of existing [Russian] documents. It will fix the output that didn't work correctly because of polyglossia: extra dot in fancy headings when `numbers = noenddot`.

> In any case, would it work for you if you could set

Since I know the workaround, I don't see a difference to use one wrong way or another. :) Yes, it will work. No changes works as well, since I know how to circumvent.

**Comment by jspitz** [comment]
> It won't change the output of existing [Russian] documents.

Yes it will. MWE:
```tex
\documentclass{scrbook}

\usepackage{polyglossia}
\usepackage{fontspec}

\setdefaultlanguage{russian}

\setmainfont{DejaVu Serif}
\newfontfamily{\cyrillicfont}{DejaVu Serif}
\setsansfont{DejaVu Sans}
\newfontfamily{\cyrillicfontsf}{DejaVu Sans}
\setmonofont{DejaVu Sans}
\newfontfamily{\cyrillicfonttt}{DejaVu Sans}

\begin{document}
\chapter{Chapter}

\section{Section}

\end{document}
```

Before your proposed change:
![grafik](https://user-images.githubusercontent.com/4538783/94260371-58b52100-ff30-11ea-96ad-c532e801e90b.png)

After:
![grafik](https://user-images.githubusercontent.com/4538783/94260582-ba758b00-ff30-11ea-838e-3e145c17af48.png)

**Comment by intractabilis** [comment]
Ah, got it! Sorry, I was wrong.

Then may I suggest a very simple solution? Set `numbers = endperiod` as default for Russian language and remove the rest of the code?

**Comment by jspitz** [comment]
Well the rest of the code is for other classes than KOMA. We need that.

And `numbers = endperiod` cannot be set locally very easily (it would require much more code than the one-liner we will have after my change for KOMA)

**Comment by intractabilis** [comment]
> Well the rest of the code is for other classes than KOMA. We need that.

Ok, the rest of the code that tweaks numbers and dots in KOMA Script.

> And `numbers = endperiod` cannot be set locally

Sad, that would be a good solution.

**Comment by jspitz** [comment]
> Ok, the rest of the code that tweaks numbers and dots in KOMA Script.

That's only one line (in the new approach):
```tex
\renewcommand*\autodot{.}%
```
The rest is not relevant (and not executed) with a KOMA class.

**Comment by intractabilis** [comment]
Thank you for spending time on fixing this and on discussion! Great job!

Closing.

**Comment by jspitz** [comment]
Thanks, but please keep opened
* I haven't merged yet (will do in a minute)
* We only close after the fix has been released.

**Comment by intractabilis** [comment]
Time to close? :)

**Comment by Lenchik** [comment]
Seems like this `autodot` solution breaks compatibility with packages that have no direct use of komascript (from users point of view).
```tex
% !TeX TS-program = lualatex
\documentclass[10pt, xcolor={dvipsnames, table, hyperref}]{beamer}
\PassOptionsToPackage{no-math}{fontspec}
\usepackage{polyglossia}
\setmainlanguage[babelshorthands=true]{russian}
\setotherlanguage{english}
\setmonofont{CMU Typewriter Text}
\newfontfamily\cyrillicfonttt{CMU Typewriter Text}
\defaultfontfeatures{Ligatures=TeX}
\setmainfont{CMU Serif}
\newfontfamily\cyrillicfont{CMU Serif}
\setsansfont{CMU Sans Serif}
\newfontfamily\cyrillicfontsf{CMU Sans Serif}

\listfiles

\begin{document}
\begin{frame}[noframenumbering,plain]
    \setcounter{framenumber}{1}
    \maketitle
\end{frame}

\begin{frame}
    \frametitle{Положения, выносимые на защиту}
    \begin{itemize}
        \item Результаты расчёта этого путём таким-то.
        \item Результаты разработки того.
        \item И ещё \dots
        \item \dots пару пунктов.
    \end{itemize}
\end{frame}
\end{document}
```
```log
! LaTeX Error: Command \autodot undefined.
```
[beamer-polyglossia-test.log](https://github.com/reutenauer/polyglossia/files/5360961/beamer-polyglossia-test.log)

If it is really this change that causes this error, can you please fix it ASAP? One "small" update and all russian presentations fail to compile...

**Comment by intractabilis** [comment]
Would defining
```LaTeX
\newcommand*\autodot{.}
```
help as a workaround while they are fixing?

**Comment by jspitz** [comment]
Fixed at 252a03a3debd.

**Comment by jspitz** [comment]
> Would defining
> 
> ```
> \newcommand*\autodot{.}
> ```
> 
> help as a workaround while they are fixing?

This would help. Or better

```tex
\providecommand*\autodot{}
```

(before loading polyglossia)

**Comment by jspitz** [comment]
I just submitted v. 1.50a to CTAN. This fixes this regression (only).
